### PR TITLE
Use test keys for keymanager and other tests

### PR DESCRIPTION
**Pull request by azdagron**
This change uses the available test keys instead of generating new keys for the keymanager tests which speeds up the tests and avoids sporadic timeout test failures we've observed in CI/CD, particularly when generating 4096 bit RSA keys.

This change also updates the testkey package to no longer fail if a test exceeds the number of pregenerated keys, instead opting to generate the new key and save it to disk so it can be included in the PR introducing the test change.

The agent fake keymanager was also updated to use test keys.

CI/CD has also been updated to fail the unit-test step if the git repository is no longer clean, implying that there were possibly test keys generated that were not part of the PR.

### CODE REVIEW

**Comment by azdagron** [comment]
Not sure if we want the change in behavior in the testkey package (now just extend the pregenerated keys instead of failing if they are exceeded). I'll let folks stew on that.

This change does mean that our keymanager tests aren't testing the actual key generation functions. This does introduce a little risk that the wiring up of those functions could be broken in production without us noticing in tests. We could write some additional unit tests to try and safeguard that. Further, integration tests should notice at least any breakage to ec256.

File: test/testkey/bucket.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+		}
+		keys = append(keys, key)
+	}
+
+	b.keys = keys
+	return nil
+}
+
+func (b *bucket[KT, K]) save() error {
+	buf := new(bytes.Buffer)
```
Comment (rturner3) [inline test/testkey/bucket.go:99]:
```suggestion
	var buf bytes.Buffer
```
nit: the zero value of `bytes.Buffer` is ready to use

**Review by rturner3 — COMMENTED** [review]
Looks good overall to me, just a couple small comments

File: test/testkey/bucket.go (hunk 1 above)
Comment (azdagron) [reply test/testkey/bucket.go:106]: I go back and forth on which pattern is more ergonomic. When using a zero valued var, you end up having to do `&buf` everywhere. Compiler escape analysis likely makes either form identical in terms of allocations. Anywho, happy to change.

File: pkg/agent/plugin/keymanager/disk/disk_test.go [hunk 2]
```diff
@@ -83,7 +84,14 @@ func TestGenerateKeyPersistence(t *testing.T) {
 func loadPlugin(t *testing.T, configFmt string, configArgs ...interface{}) (keymanager.KeyManager, error) {
 	km := new(keymanager.V1)
 	var configErr error
-	plugintest.Load(t, disk.BuiltIn(), km,
+
+	var keys testkey.Keys
+	plugintest.Load(t, disk.AsBuiltIn(disk.New(disk.Generators{
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/plugin/keymanager/disk/disk_test.go:89]: My recent refactor should make this better. Please let me know if it's still hard to read.

**Comment by azdagron** [comment]
Enough changed in my latest refactor that most of the code outside of the testkey package changed significantly. I thought it would be ok to squash and rebase. Hope that doesn't cause any hardship.

File: pkg/server/plugin/keymanager/base/keymanagerbase.go [hunk 3]
```diff
@@ -26,42 +26,40 @@ type KeyEntry struct {
...
-	WriteEntries       func(ctx context.Context, entries []*KeyEntry) error
-	GenerateRSA2048Key func() (*rsa.PrivateKey, error)
-	GenerateRSA4096Key func() (*rsa.PrivateKey, error)
-	GenerateEC256Key   func() (*ecdsa.PrivateKey, error)
-	GenerateEC384Key   func() (*ecdsa.PrivateKey, error)
+type Config struct {
+	// Generator is an optional key generator.
+	Generator Generator
+
+	// WriteEntries is an optional callback used to persisted key entries
```
Comment (rturner3) [inline pkg/server/plugin/keymanager/base/keymanagerbase.go:35]:
```suggestion
	// WriteEntries is an optional callback used to persist key entries
```

File: pkg/agent/plugin/keymanager/base/keymanagerbase.go [hunk 4]
```diff
@@ -26,42 +26,40 @@ type KeyEntry struct {
...
-	WriteEntries       func(ctx context.Context, allEntries []*KeyEntry, newEntry *KeyEntry) error
-	GenerateRSA2048Key func() (*rsa.PrivateKey, error)
-	GenerateRSA4096Key func() (*rsa.PrivateKey, error)
-	GenerateEC256Key   func() (*ecdsa.PrivateKey, error)
-	GenerateEC384Key   func() (*ecdsa.PrivateKey, error)
+type Config struct {
+	// Generator is an optional key generator.
+	Generator Generator
+
+	// WriteEntries is an optional callback used to persisted key entries
```
Comment (rturner3) [inline pkg/agent/plugin/keymanager/base/keymanagerbase.go:35]:
```suggestion
	// WriteEntries is an optional callback used to persist key entries
```

File: test/testkey/bucket.go (hunk 1 above)
Comment (rturner3) [reply test/testkey/bucket.go]: Yeah, they are basically equivalent. I tend to see the zero valued var usage more often and in Go documentation examples, so I generally lean towards that form.

**Review by rturner3 — APPROVED** [review]
LGTM! I noticed a couple minor typos, but it looks good as is to merge if you want to get this wrapped up before the next release.

**Comment by azdagron** [comment]
We're in no hurry here. I'll fix the typos. Thanks for the review!

**Review by MarcosDY — APPROVED** [review]
LGTM!!!

**Pull request merged**